File: INLBreadcrumbs/INLBreadcrumbManager.m
//
//  INLBreadcrumbService.m
//  INLBreadcrumbs
//
//  Created by Tomas Hakel on 02/12/2015.
//  Copyright © 2015 Inloop, s.r.o. See LICENSE.md
//

#import "INLBreadcrumbManager.h"
#import "INLBreadcrumb.h"
#import "INLBreadcrumbCompatibleController.h"

@interface INLBreadcrumbManager ()

@property (strong, nonatomic) NSMutableArray<INLBreadcrumb *> * breadcrumbStack;

@end


@implementation INLBreadcrumbManager

-(instancetype)init {
	if (self = [super init]) {
		self.breadcrumbStack = [@[] mutableCopy];
		self.cancelButtonTitle = nil;
		self.breadcrumbIndicator = nil;
	}
	return self;
}

-(NSArray<INLBreadcrumb *> *)breadcrumbs {
	return self.breadcrumbStack;
}

-(void)pushBreadcrumb:(INLBreadcrumb *)breadcrumb {
	[self.breadcrumbStack addObject:breadcrumb];
}

-(void)popBreadcrumb:(INLBreadcrumb *)breadcrumb {
	[self.breadcrumbStack removeLastObject];
}

+(instancetype)defaultManager {
	static INLBreadcrumbManager * defaultManager = nil;
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		defaultManager = [[INLBreadcrumbManager alloc] init];
	});
	return defaultManager;
}

+(instancetype)managerForKey:(NSString *)key {
	static NSMutableDictionary<NSString *, INLBreadcrumbManager *> * managers = nil;
	if (!managers) { managers = [@{} mutableCopy]; }
	
	INLBreadcrumbManager * manager = managers[key];
	if (!manager) {
		manager = [[INLBreadcrumbManager alloc] init];
		managers[key] = manager;
	}
	return manager;
}

@end

File: INLBreadcrumbs/INLBreadcrumbManager.swift
//
//  INLBreadcrumbManager.swift
//  INLBreadcrumbs
//
//  Created by Petr Tomicek on 13.12.16.
//  Copyright © 2016 Inloop, s.r.o. All rights reserved.
//

import Foundation

public class INLBreadcrumbManager
{
    var cancelButtonTitle: String?
    var breadcrumbIndicator: String?
    
    static let shared = INLBreadcrumbManager()
    
    private var managers: [String:INLBreadcrumbManager]
    
    private var breadcrumbStack: [INLBreadcrumb]
    
    public init() {
        
        self.breadcrumbStack = []
        self.managers = [:]
    }
    
    public func managerForKey(key: String) ->  INLBreadcrumbManager {
        
        if let manager = managers[key] {
            
            return manager
        }
        else {
            
            let manager = INLBreadcrumbManager()
            managers[key] = manager
            return manager
        }
    }
    
    public func pushBreadcrumb(breadcrumb: INLBreadcrumb) {
        
        self.breadcrumbStack.append(breadcrumb)
    }
    
    public func popBreadcrumb(breadcrumb: INLBreadcrumb) {
        
        self.breadcrumbStack.removeLast()
    }
    
    public func breadcrumbs() -> [INLBreadcrumb] {
        
        return self.breadcrumbStack
    }
}
